keps/sig-api-machinery/20190802-dynamic-coordinated-storage-version.md

```diff
@@ -0,0 +1,289 @@
+   * [Garbage collection of the StorageVersion objects of removed resources](#garbage-collection-of-the-storageversion-objects-of-removed-resources)
+   * [Using the selected storage version when serializing data for etcd](#using-the-selected-storage-version-when-serializing-data-for-etcd)
+   * [Updating the discovery document](#updating-the-discovery-document)
+   * [CRDs](#crds)
+   * [Aggregated API servers](#aggregated-api-servers)
+* [Backwards Compatibility](#backwards-compatibility)
+* [Graduation Plan](#graduation-plan)
+* [FAQ](#faq)
+* [Future work](#future-work)
+* [Alternatives](#alternatives)
+   * [Exposing if API servers in a HA setup are using the same storage version](#exposing-if-api-servers-in-a-ha-setup-are-using-the-same-storage-version)
+   * [Letting the storage migrator detect if API server instances are in agreement](#letting-the-storage-migrator-detect-if-api-server-instances-are-in-agreement)
+
+## Overview
+
+During the rolling update of an HA master, API server instances *i)* use
+different storage versions for a built-in API resource, and *ii)* show different
+storageVersionHash in the discovery document. These fact make the storage
+migrator not working properly ([details][]).
+
+[storageVersionHash]:https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go#L979
+[details]:https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/35-storage-version-hash.md#ha-masters
+
+We propose a mechanism for HA API servers to vote on the storage version to use,
+then show the selected storage version in the discovery document and switch to
+use the selected version atomically.
+
+## API changes
+
+### Storage Version API
+
+We introduce a new API `StorageVersion`, in a new API group
+`kube-apiserver.internal.k8s.io/v1alpha1`. 
+
+```golang
+// Storage version for a specific resource.
+type StorageVersion struct {
+  TypeMeta
+  // The name is the GVR, i.e., <group>.<version>.<resource>.
+  ObjectMeta
+  // The selected storage version. All API server intances encode the resource
+  // objects in this version when committing it to etcd.
+  Selectedversion string
+  // Proposed storage version candidate, keyed by the ID of the proposing API
+  // server.
+  CandidateVersions []VersionCandidate
+}
+
+type VersionCandidate struct {
+  // The ID of the reporting API server. 
+  APIServerID string
+  // The preferred storage version of the reporting API server. The preferred
+  // storage version is hardcoded in each Kubernetes release.
+  PreferredVersion string
+}
+```
+
+We will discuss how the `StorageVersion` is updated and used in the [Changes to API
+servers](#changes-to-api-servers) section.
+
+## Changes to API servers
+
+### Curating a list of participating API servers in HA master
+
+Currently, API servers maintain such a list in the "kubernetes" endpoints. See
```

Comment:
We have a way to do that - this doesn't mean we always do.
For example in GKE, this endpoint contains only IP of the LB in front of apiservers (whether that's good or bad is a separate discussion, but I guess we may not be the only one doing something like that).

Reply:
Other operators might have other forms of deployments that also hide the individual apiserver endpoints, so we can't rely on the current "kubernetes" service endpoints to get a reliable list of apiservers.

As I mentioned in the KEP, I will send another KEP to formalize the procedure of apiserver census. Operators will need to configure an ID for each apiserver to make it work.